types/styled-components/index.d.ts

```diff
@@ -190,6 +191,7 @@ export interface StyledComponentBase<
              * String types need to be cast to themselves to become literal types (as={'a' as 'a'}).
              */
             as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
+            forwardedAs?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
```

Comment:
I guess this assumes that `as` takes a component by convention. Shouldn't this actually infer the type of `as` from the child component?